### PR TITLE
Reify integration problems

### LINKED ISSUE #533: Reify integration problems
Structs for ODEs, IVPs, step size controllers would make life easier when passing arguments to integrators, and would relate the integrators to each other (by generalizing the ODEs).

### PR DESCRIPTION
**Pull request by eggrobin**
This fixes #533.
It also moves `DoublePrecision` somewhere saner.

### CODE REVIEW

File: integrators/embedded_explicit_runge_kutta_nyström_integrator_body.hpp [hunk 1]
```diff
@@ -70,50 +70,49 @@ inline EmbeddedExplicitRungeKuttaNyströmIntegrator::
...
+  using Acceleration = typename ODE<Position>::Acceleration;
 
   // Argument checks.
-  int const dimension = initial_value.positions.size();
-  CHECK_EQ(initial_value.momenta.size(), dimension);
-  CHECK_NE(first_time_step, Time());
-  Sign const integration_direction = Sign(first_time_step);
+  CHECK_NOTNULL(problem.initial_state);
+  int const dimension = problem.initial_state->positions.size();
+  CHECK_EQ(problem.initial_state->velocities.size(), dimension);
```
Comment (pleroy) [inline integrators/embedded_explicit_runge_kutta_nyström_integrator_body.hpp]: Expected value (i.e., `dimension`) should go first.

File: integrators/embedded_explicit_runge_kutta_nyström_integrator_body.hpp [hunk 2]
```diff
@@ -70,50 +70,49 @@ inline EmbeddedExplicitRungeKuttaNyströmIntegrator::
...
+  CHECK_NE(adaptive_step_size.first_time_step, Time());
```
Comment (pleroy) [inline integrators/embedded_explicit_runge_kutta_nyström_integrator_body.hpp]: Same.

File: integrators/embedded_explicit_runge_kutta_nyström_integrator_body.hpp [hunk 3]
```diff
@@ -70,50 +70,49 @@ inline EmbeddedExplicitRungeKuttaNyströmIntegrator::
...
   // Current velocity.
-  std::vector<DoublePrecision<Velocity>> v_hat = initial_value.momenta;
+  std::vector<DoublePrecision<Velocity>>& v_hat = current_state.velocities;
 
-  // Difference between the low- and high-order approximations of the position.
-  std::vector<Displacement> q_error_estimate(dimension);
-  // Difference between the low- and high-order approximations of the velocity.
-  std::vector<Velocity> v_error_estimate(dimension);
+  // Difference between the low- and high-order approximations.
+ typename ODE<Position>::SystemStateError error_estimate;
```
Comment (pleroy) [inline integrators/embedded_explicit_runge_kutta_nyström_integrator_body.hpp]: Bad indent.

File: integrators/embedded_explicit_runge_kutta_nyström_integrator_body.hpp [hunk 4]
```diff
@@ -70,50 +70,49 @@ inline EmbeddedExplicitRungeKuttaNyströmIntegrator::
...
-  DoublePrecision<Time> t = initial_value.time;
+  DoublePrecision<Instant>& t = current_state.time;
 
   // Position increment (high-order).
   std::vector<Displacement> ∆q_hat(dimension);
   // Velocity increment (high-order).
   std::vector<Velocity> ∆v_hat(dimension);
   // Current position.
-  std::vector<DoublePrecision<Position>> q_hat = initial_value.positions;
+  std::vector<DoublePrecision<Position>>& q_hat = current_state.positions;
```
Comment (pleroy) [inline integrators/embedded_explicit_runge_kutta_nyström_integrator_body.hpp]: Non-const references are fine but unusual.  Comment that this is really what you want.

Comment (eggrobin) [inline integrators/embedded_explicit_runge_kutta_nyström_integrator_body.hpp]: Done.

File: integrators/embedded_explicit_runge_kutta_nyström_integrator_body.hpp (hunk 3 above)
Comment (eggrobin) [inline integrators/embedded_explicit_runge_kutta_nyström_integrator_body.hpp]: We really need to get that linter working on files with unicode in the name.

File: integrators/embedded_explicit_runge_kutta_nyström_integrator_body.hpp (hunk 2 above)
Comment (eggrobin) [inline integrators/embedded_explicit_runge_kutta_nyström_integrator_body.hpp]: Done.

File: integrators/embedded_explicit_runge_kutta_nyström_integrator_body.hpp (hunk 1 above)
Comment (eggrobin) [inline integrators/embedded_explicit_runge_kutta_nyström_integrator_body.hpp]: Done.

**Pull request merged**